File: src/java1/algorithms/binary/CountingBits.java
// Dynamic programming pattern:- TC:O(n) SC:O(n)
package java1.algorithms.binary;

import java.util.Arrays;

public class CountingBits {
    private static int[] countingBits(int n) {
        int[] dp = new int[n+1];
        dp[0] = 0;
        int offset = 1;
        for(int i=1; i< n+1; i++) {
            if(offset * 2 == i) {
                offset = i;
            }
            dp[i] = 1+ dp[i-offset];
        }
        return dp;
    }

    public static void main(String[] args) {
        int num = 3;
        System.err.println(Arrays.toString(countingBits(num)));
    }
}

File: src/javascript/algorithms/binary/countingBits.js
// Dynamic programming:- TC:O(n) SC:O(n)

function countingBits(num) {
    let dp = new Array(num+1).fill(0);
    let offset = 1; 
    for(let i=1; i< num+1; i++) {
        if(offset * 2 === i) {
            offset = i;
        }
        dp[i] = 1 +dp[i-offset];
    }
    return dp;
}

let num = 3;
console.log(countingBits(num));